File: test-automation-copilot/copilot-core/pkg/parser/java_parser.go
package parser

import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	sitter "github.com/smacker/go-tree-sitter"
	"github.com/smacker/go-tree-sitter/java"
)

type JavaParser struct {
	parser *sitter.Parser
}

type ParsedClass struct {
	ClassName   string        `json:"className"`
	FilePath    string        `json:"filePath"`
	Package     string        `json:"package"`
	Imports     []string      `json:"imports"`
	Fields      []FieldInfo   `json:"fields"`
	Methods     []MethodInfo  `json:"methods"`
	Annotations []string      `json:"annotations"`
	SuperClass  string        `json:"superClass,omitempty"`
	Type        string        `json:"type"` // pageObject, test, step, utility, unknown
}

type FieldInfo struct {
	Name         string   `json:"name"`
	Type         string   `json:"type"`
	LocatorType  string   `json:"locatorType,omitempty"`
	LocatorValue string   `json:"locatorValue,omitempty"`
	Annotations  []string `json:"annotations"`
}

type MethodInfo struct {
	Name        string          `json:"name"`
	ReturnType  string          `json:"returnType"`
	Parameters  []ParameterInfo `json:"parameters"`
	Annotations []string        `json:"annotations"`
	Body        string          `json:"body"`
}

type ParameterInfo struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

func NewJavaParser() *JavaParser {
	parser := sitter.NewParser()
	parser.SetLanguage(java.GetLanguage())

	return &JavaParser{
		parser: parser,
	}
}

func (p *JavaParser) ParseFile(filePath string) (*ParsedClass, error) {
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	tree, err := p.parser.ParseCtx(context.Background(), nil, content)
	if err != nil {
		return nil, fmt.Errorf("failed to parse: %w", err)
	}

	result := p.extractClassInfo(tree.RootNode(), content, filePath)
	return result, nil
}

func (p *JavaParser) ParseDirectory(dirPath string) ([]*ParsedClass, error) {
	var results []*ParsedClass

	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !info.IsDir() && strings.HasSuffix(path, ".java") {
			parsed, err := p.ParseFile(path)
			if err != nil {
				fmt.Printf("Warning: failed to parse %s: %v\n", path, err)
				return nil // Continue with other files
			}
			results = append(results, parsed)
		}

		return nil
	})

	return results, err
}

func (p *JavaParser) extractClassInfo(node *sitter.Node, content []byte, filePath string) *ParsedClass {
	result := &ParsedClass{
		FilePath:    filePath,
		Imports:     []string{},
		Fields:      []FieldInfo{},
		Methods:     []MethodInfo{},
		Annotations: []string{},
	}

	// Extract package
	packageNode := findNode(node, "package_declaration")
	if packageNode != nil {
		pkgText := nodeText(packageNode, content)
		result.Package = strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(pkgText, "package"), ";"))
	}

	// Extract imports
	importNodes := findAllNodes(node, "import_declaration")
	for _, importNode := range importNodes {
		importText := nodeText(importNode, content)
		importText = strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(importText, "import"), ";"))
		result.Imports = append(result.Imports, importText)
	}

	// Extract class declaration
	classNode := findNode(node, "class_declaration")
	if classNode != nil {
		// Class name
		identifierNode := findNode(classNode, "identifier")
		if identifierNode != nil {
			result.ClassName = nodeText(identifierNode, content)
		}

		// Annotations
		result.Annotations = extractAnnotations(classNode, content)

		// Super class
		superclassNode := findNode(classNode, "superclass")
		if superclassNode != nil {
			typeNode := findNode(superclassNode, "type_identifier")
			if typeNode != nil {
				result.SuperClass = nodeText(typeNode, content)
			}
		}

		// Class body
		classBody := findNode(classNode, "class_body")
		if classBody != nil {
			result.Fields = extractFields(classBody, content)
			result.Methods = extractMethods(classBody, content)
		}
	}

	// Determine class type
	result.Type = determineClassType(result)

	return result
}

func extractFields(classBody *sitter.Node, content []byte) []FieldInfo {
	var fields []FieldInfo
	fieldNodes := findAllNodes(classBody, "field_declaration")

	for _, fieldNode := range fieldNodes {
		typeNode := findNode(fieldNode, "type")
		declaratorNode := findNode(fieldNode, "variable_declarator")

		if typeNode != nil && declaratorNode != nil {
			identifierNode := findNode(declaratorNode, "identifier")
			if identifierNode != nil {
				field := FieldInfo{
					Name:        nodeText(identifierNode, content),
					Type:        nodeText(typeNode, content),
					Annotations: extractAnnotations(fieldNode, content),
				}

				// Extract locator information
				locatorType, locatorValue := extractLocatorInfo(fieldNode, content)
				if locatorType != "" {
					field.LocatorType = locatorType
					field.LocatorValue = locatorValue
				}

				fields = append(fields, field)
			}
		}
	}

	return fields
}

func extractMethods(classBody *sitter.Node, content []byte) []MethodInfo {
	var methods []MethodInfo
	methodNodes := findAllNodes(classBody, "method_declaration")

	for _, methodNode := range methodNodes {
		nameNode := findNode(methodNode, "identifier")
		var typeNode *sitter.Node
		if tn := findNode(methodNode, "type"); tn != nil {
			typeNode = tn
		} else if vt := findNode(methodNode, "void_type"); vt != nil {
			typeNode = vt
		}

		if nameNode != nil && typeNode != nil {
			method := MethodInfo{
				Name:        nodeText(nameNode, content),
				ReturnType:  nodeText(typeNode, content),
				Parameters:  extractParameters(methodNode, content),
				Annotations: extractAnnotations(methodNode, content),
				Body:        getMethodBody(methodNode, content),
			}
			methods = append(methods, method)
		}
	}

	return methods
}

func extractParameters(methodNode *sitter.Node, content []byte) []ParameterInfo {
	var parameters []ParameterInfo
	paramsNode := findNode(methodNode, "formal_parameters")

	if paramsNode != nil {
		paramNodes := findAllNodes(paramsNode, "formal_parameter")
		for _, paramNode := range paramNodes {
			typeNode := findNode(paramNode, "type")
			identifierNode := findNode(paramNode, "identifier")

			if typeNode != nil && identifierNode != nil {
				parameters = append(parameters, ParameterInfo{
					Name: nodeText(identifierNode, content),
					Type: nodeText(typeNode, content),
				})
			}
		}
	}

	return parameters
}

func extractAnnotations(node *sitter.Node, content []byte) []string {
	var annotations []string

	for i := 0; i < int(node.ChildCount()); i++ {
		child := node.Child(i)
		if child.Type() == "modifiers" {
			annotationNodes := findAllNodes(child, "annotation")
			for _, annotNode := range annotationNodes {
				annotations = append(annotations, nodeText(annotNode, content))
			}
		} else if child.Type() == "annotation" {
			annotations = append(annotations, nodeText(child, content))
		}
	}

	return annotations
}

func extractLocatorInfo(fieldNode *sitter.Node, content []byte) (string, string) {
	fieldText := nodeText(fieldNode, content)

	// Check for @FindBy annotation
	if strings.Contains(fieldText, "@FindBy") {
		// Pattern: @FindBy(id = "username")
		for _, locType := range []string{"id", "name", "css", "xpath", "className", "tagName", "linkText", "partialLinkText"} {
			pattern := fmt.Sprintf(`%s\s*=\s*"([^"]+)"`, locType)
			if idx := strings.Index(fieldText, locType+" ="); idx != -1 {
				start := strings.Index(fieldText[idx:], `"`) + idx + 1
				end := strings.Index(fieldText[start:], `"`) + start
				if start > idx && end > start {
					return fmt.Sprintf("@FindBy(%s)", locType), fieldText[start:end]
				}
			}
		}
	}

	// Check for By.xxx() pattern
	if strings.Contains(fieldText, "By.") {
		for _, locType := range []string{"id", "name", "cssSelector", "xpath", "className", "tagName", "linkText", "partialLinkText"} {
			pattern := fmt.Sprintf(`By\.%s\("([^"]+)"\)`, locType)
			if strings.Contains(fieldText, "By."+locType) {
				start := strings.Index(fieldText, "By."+locType+"(\"") + len("By."+locType+"(\"")
				end := strings.Index(fieldText[start:], `"`) + start
				if end > start {
					return fmt.Sprintf("By.%s", locType), fieldText[start:end]
				}
			}
		}
	}

	return "", ""
}

func getMethodBody(methodNode *sitter.Node, content []byte) string {
	bodyNode := findNode(methodNode, "block")
	if bodyNode != nil {
		return nodeText(bodyNode, content)
	}
	return ""
}

func determineClassType(class *ParsedClass) string {
	className := strings.ToLower(class.ClassName)
	filePath := strings.ToLower(class.FilePath)

	// Check for Page Object
	if strings.Contains(className, "page") || strings.Contains(filePath, "/pages/") {
		return "pageObject"
	}

	// Check for Test class
	if strings.Contains(className, "test") || strings.Contains(filePath, "/tests/") || strings.Contains(filePath, "/test/") {
		return "test"
	}

	// Check for Step definition (Cucumber)
	if strings.Contains(className, "step") || strings.Contains(filePath, "/steps/") {
		return "step"
	}
	for _, ann := range class.Annotations {
		if strings.Contains(ann, "@Given") || strings.Contains(ann, "@When") || strings.Contains(ann, "@Then") {
			return "step"
		}
	}

	// Check for utility class
	if strings.Contains(className, "util") ||
		strings.Contains(className, "helper") ||
		strings.Contains(className, "factory") ||
		strings.Contains(filePath, "/utils/") ||
		strings.Contains(filePath, "/utilities/") {
		return "utility"
	}

	return "unknown"
}

// Helper functions
func findNode(node *sitter.Node, nodeType string) *sitter.Node {
	if node.Type() == nodeType {
		return node
	}

	for i := 0; i < int(node.ChildCount()); i++ {
		if found := findNode(node.Child(i), nodeType); found != nil {
			return found
		}
	}

	return nil
}

func findAllNodes(node *sitter.Node, nodeType string) []*sitter.Node {
	var results []*sitter.Node

	if node.Type() == nodeType {
		results = append(results, node)
	}

	for i := 0; i < int(node.ChildCount()); i++ {
		results = append(results, findAllNodes(node.Child(i), nodeType)...)
	}

	return results
}

func nodeText(node *sitter.Node, content []byte) string {
	return string(content[node.StartByte():node.EndByte()])
}

File: test-automation-copilot/src/core/parsers/JavaParser.ts
import * as fs from 'fs';
import * as path from 'path';
import Parser from 'tree-sitter';
import Java from 'tree-sitter-java';

export interface ParsedClass {
    className: string;
    filePath: string;
    package: string;
    imports: string[];
    fields: FieldInfo[];
    methods: MethodInfo[];
    annotations: string[];
    superClass?: string;
    type: 'pageObject' | 'test' | 'step' | 'utility' | 'unknown';
}

export interface FieldInfo {
    name: string;
    type: string;
    locatorType?: string; // By.id, @FindBy, etc.
    locatorValue?: string;
    annotations: string[];
}

export interface MethodInfo {
    name: string;
    returnType: string;
    parameters: ParameterInfo[];
    annotations: string[];
    body: string;
}

export interface ParameterInfo {
    name: string;
    type: string;
}

export class JavaParser {
    private parser: Parser;

    constructor() {
        this.parser = new Parser();
        this.parser.setLanguage(Java);
    }

    /**
     * Parse a single Java file
     */
    public parseFile(filePath: string): ParsedClass | null {
        try {
            const content = fs.readFileSync(filePath, 'utf8');
            const tree = this.parser.parse(content);

            return this.extractClassInfo(tree.rootNode, content, filePath);
        } catch (error) {
            console.error(`Error parsing ${filePath}:`, error);
            return null;
        }
    }

    /**
     * Parse all Java files in a directory
     */
    public parseDirectory(dirPath: string): ParsedClass[] {
        const results: ParsedClass[] = [];

        const walk = (dir: string) => {
            const files = fs.readdirSync(dir);

            for (const file of files) {
                const filePath = path.join(dir, file);
                const stat = fs.statSync(filePath);

                if (stat.isDirectory()) {
                    walk(filePath);
                } else if (file.endsWith('.java')) {
                    const parsed = this.parseFile(filePath);
                    if (parsed) {
                        results.push(parsed);
                    }
                }
            }
        };

        walk(dirPath);
        return results;
    }

    private extractClassInfo(node: Parser.SyntaxNode, content: string, filePath: string): ParsedClass {
        const result: ParsedClass = {
            className: '',
            filePath,
            package: '',
            imports: [],
            fields: [],
            methods: [],
            annotations: [],
            type: 'unknown'
        };

        // Extract package
        const packageNode = this.findNode(node, 'package_declaration');
        if (packageNode) {
            result.package = this.getNodeText(packageNode, content).replace('package ', '').replace(';', '').trim();
        }

        // Extract imports
        const importNodes = this.findAllNodes(node, 'import_declaration');
        result.imports = importNodes.map(n =>
            this.getNodeText(n, content).replace('import ', '').replace(';', '').trim()
        );

        // Extract class declaration
        const classNode = this.findNode(node, 'class_declaration');
        if (classNode) {
            // Class name
            const identifierNode = this.findNode(classNode, 'identifier');
            if (identifierNode) {
                result.className = this.getNodeText(identifierNode, content);
            }

            // Annotations
            result.annotations = this.extractAnnotations(classNode, content);

            // Super class
            const superclassNode = this.findNode(classNode, 'superclass');
            if (superclassNode) {
                const typeNode = this.findNode(superclassNode, 'type_identifier');
                if (typeNode) {
                    result.superClass = this.getNodeText(typeNode, content);
                }
            }

            // Class body
            const classBody = this.findNode(classNode, 'class_body');
            if (classBody) {
                // Extract fields
                result.fields = this.extractFields(classBody, content);

                // Extract methods
                result.methods = this.extractMethods(classBody, content);
            }
        }

        // Determine class type
        result.type = this.determineClassType(result);

        return result;
    }

    private extractFields(classBody: Parser.SyntaxNode, content: string): FieldInfo[] {
        const fields: FieldInfo[] = [];
        const fieldNodes = this.findAllNodes(classBody, 'field_declaration');

        for (const fieldNode of fieldNodes) {
            const typeNode = this.findNode(fieldNode, 'type');
            const declaratorNode = this.findNode(fieldNode, 'variable_declarator');

            if (typeNode && declaratorNode) {
                const identifierNode = this.findNode(declaratorNode, 'identifier');

                if (identifierNode) {
                    const field: FieldInfo = {
                        name: this.getNodeText(identifierNode, content),
                        type: this.getNodeText(typeNode, content),
                        annotations: this.extractAnnotations(fieldNode, content)
                    };

                    // Extract locator information
                    const locatorInfo = this.extractLocatorInfo(fieldNode, content);
                    if (locatorInfo) {
                        field.locatorType = locatorInfo.type;
                        field.locatorValue = locatorInfo.value;
                    }

                    fields.push(field);
                }
            }
        }

        return fields;
    }

    private extractMethods(classBody: Parser.SyntaxNode, content: string): MethodInfo[] {
        const methods: MethodInfo[] = [];
        const methodNodes = this.findAllNodes(classBody, 'method_declaration');

        for (const methodNode of methodNodes) {
            const nameNode = this.findNode(methodNode, 'identifier');
            const typeNode = this.findNode(methodNode, 'type') || this.findNode(methodNode, 'void_type');

            if (nameNode && typeNode) {
                const method: MethodInfo = {
                    name: this.getNodeText(nameNode, content),
                    returnType: this.getNodeText(typeNode, content),
                    parameters: this.extractParameters(methodNode, content),
                    annotations: this.extractAnnotations(methodNode, content),
                    body: this.getMethodBody(methodNode, content)
                };

                methods.push(method);
            }
        }

        return methods;
    }

    private extractParameters(methodNode: Parser.SyntaxNode, content: string): ParameterInfo[] {
        const parameters: ParameterInfo[] = [];
        const paramsNode = this.findNode(methodNode, 'formal_parameters');

        if (paramsNode) {
            const paramNodes = this.findAllNodes(paramsNode, 'formal_parameter');

            for (const paramNode of paramNodes) {
                const typeNode = this.findNode(paramNode, 'type');
                const identifierNode = this.findNode(paramNode, 'identifier');

                if (typeNode && identifierNode) {
                    parameters.push({
                        name: this.getNodeText(identifierNode, content),
                        type: this.getNodeText(typeNode, content)
                    });
                }
            }
        }

        return parameters;
    }

    private extractAnnotations(node: Parser.SyntaxNode, content: string): string[] {
        const annotations: string[] = [];

        for (const child of node.children) {
            if (child.type === 'modifiers') {
                const annotationNodes = this.findAllNodes(child, 'annotation');
                annotations.push(...annotationNodes.map(n => this.getNodeText(n, content)));
            } else if (child.type === 'annotation') {
                annotations.push(this.getNodeText(child, content));
            }
        }

        return annotations;
    }

    private extractLocatorInfo(fieldNode: Parser.SyntaxNode, content: string): { type: string, value: string } | null {
        const fieldText = this.getNodeText(fieldNode, content);

        // Check for @FindBy annotation
        const findByMatch = fieldText.match(/@FindBy\((.*?)\)/);
        if (findByMatch) {
            const params = findByMatch[1];
            const typeMatch = params.match(/(id|name|css|xpath|className|tagName|linkText|partialLinkText)\s*=\s*"([^"]+)"/);
            if (typeMatch) {
                return { type: `@FindBy(${typeMatch[1]})`, value: typeMatch[2] };
            }
        }

        // Check for By.xxx() pattern
        const byMatch = fieldText.match(/By\.(id|name|css|xpath|className|tagName|linkText|partialLinkText)\("([^"]+)"\)/);
        if (byMatch) {
            return { type: `By.${byMatch[1]}`, value: byMatch[2] };
        }

        return null;
    }

    private getMethodBody(methodNode: Parser.SyntaxNode, content: string): string {
        const bodyNode = this.findNode(methodNode, 'block');
        return bodyNode ? this.getNodeText(bodyNode, content) : '';
    }

    private determineClassType(classInfo: ParsedClass): 'pageObject' | 'test' | 'step' | 'utility' | 'unknown' {
        const className = classInfo.className.toLowerCase();
        const filePath = classInfo.filePath.toLowerCase();

        // Check for Page Object
        if (className.includes('page') || filePath.includes('/pages/')) {
            return 'pageObject';
        }

        // Check for Test class
        if (className.includes('test') || filePath.includes('/tests/') || filePath.includes('/test/')) {
            return 'test';
        }

        // Check for Step definition (Cucumber)
        if (className.includes('step') ||
            filePath.includes('/steps/') ||
            classInfo.annotations.some(a => a.includes('@Given') || a.includes('@When') || a.includes('@Then'))) {
            return 'step';
        }

        // Check for utility class
        if (className.includes('util') ||
            className.includes('helper') ||
            className.includes('factory') ||
            filePath.includes('/utils/') ||
            filePath.includes('/utilities/')) {
            return 'utility';
        }

        return 'unknown';
    }

    private findNode(node: Parser.SyntaxNode, type: string): Parser.SyntaxNode | null {
        if (node.type === type) {
            return node;
        }

        for (const child of node.children) {
            const found = this.findNode(child, type);
            if (found) {
                return found;
            }
        }

        return null;
    }

    private findAllNodes(node: Parser.SyntaxNode, type: string): Parser.SyntaxNode[] {
        const results: Parser.SyntaxNode[] = [];

        if (node.type === type) {
            results.push(node);
        }

        for (const child of node.children) {
            results.push(...this.findAllNodes(child, type));
        }

        return results;
    }

    private getNodeText(node: Parser.SyntaxNode, content: string): string {
        return content.substring(node.startIndex, node.endIndex);
    }
}
